tests/unit/s2n_fork_generation_number_test.c

```diff
@@ -0,0 +1,349 @@
+    char process_child_stack[PROCESS_CHILD_STACK_SIZE];
+    EXPECT_NOT_NULL(process_child_stack);
+
+    int proc_pid = clone(s2n_unit_test_clone_child_process, (void *) (process_child_stack + PROCESS_CHILD_STACK_SIZE), 0, (void *) &return_fork_generation_number);
+    EXPECT_NOT_EQUAL(proc_pid, -1);
+
+    s2n_verify_child_exit_status(proc_pid);
+
+    /* Verify stability */
+    return_fork_generation_number = UNEXPECTED_RETURNED_FGN;
+    EXPECT_OK(s2n_get_fork_generation_number(&return_fork_generation_number));
+    EXPECT_EQUAL(return_fork_generation_number, parent_process_fgn);
+#endif
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_unit_tests_common(struct fgn_test_case *test_case)
+{
+    uint64_t return_fork_generation_number = 0;
+
+    EXPECT_OK(s2n_get_fork_generation_number(&return_fork_generation_number));
+    EXPECT_EQUAL(return_fork_generation_number, 0);
+
+    /* Should be idempotent if no fork event occurred */
+    return_fork_generation_number = UNEXPECTED_RETURNED_FGN;
+    EXPECT_OK(s2n_get_fork_generation_number(&return_fork_generation_number));
+    EXPECT_EQUAL(return_fork_generation_number, 0);
+
+    /* Should be idempotent in threaded environment as well */
+    EXPECT_EQUAL(s2n_unit_test_thread(return_fork_generation_number), S2N_SUCCESS);
+
+    /* Cached FGN should increment if a fork event occurs */
+    EXPECT_EQUAL(s2n_unit_test_fork(return_fork_generation_number, FORK_LEVEL_FOR_TESTS), S2N_SUCCESS);
+    EXPECT_EQUAL(s2n_unit_test_fork_check_threads_first(return_fork_generation_number), S2N_SUCCESS);
+
+    /* Some fork detection mechanisms can also detect forks through clone() */
+    if (test_case->test_case_must_pass_clone_test == CLONE_TEST_YES) {
+        EXPECT_EQUAL(s2n_assert_madv_wipeonfork_is_supported() || s2n_assert_map_inherit_zero_is_supported(), true);
+        EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+    }
+    else if (test_case->test_case_must_pass_clone_test == CLONE_TEST_DETERMINE_AT_RUNTIME) {
+        if (s2n_assert_madv_wipeonfork_is_supported() == true ||
+            s2n_assert_map_inherit_zero_is_supported() == true) {
+            EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_default_cb(struct fgn_test_case *test_case)
+{
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_pthread_atfork_cb(struct fgn_test_case *test_case)
+{
+    POSIX_GUARD_RESULT(s2n_ignore_wipeonfork_and_inherit_zero_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_madv_wipeonfork_cb(struct fgn_test_case *test_case)
+{
+    if (s2n_assert_madv_wipeonfork_is_supported() == false) {
+        FGN_TEST_CASE_PRINT_MSG_INFO("Test case not supported. Skipping.", test_case);
+        return S2N_SUCCESS;
+    }
```

Comment:
Won't this lead to a LOT of messages?

Reply:
Our tests shouldn't print anything on a success, and one per run of s2n_fork_generation_number_test seems like a lot even for a debug message. Why would you want it to print more than once?

s2n-tls has a mechanism for frustrated engineers debugging. We have a "S2N_TEST_DEBUG" flag you can set to get more verbose output. See https://github.com/aws/s2n-tls/blob/b3142598af1d5031fa245cecd8e4d3ac4f6a4c38/tests/s2n_test.h#L220-L227.